File: leetcode/RemoveDuplicatesFromSortedArray/removeDuplicates.c
#include <stdio.h>

int removeDuplicates(int a[], int array_size);

// The main function
int main(){

	// Different test cases..
	int my_array1[] = {1, 1, 2, 3, 5, 6, 6, 7, 10, 25, 100, 123, 123};
	int my_array1_size = 13;

	int my_array2[] = {1, 2, 3, 5, 6};
	int my_array2_size = 5;

	int my_array3[] = {1, 1, 1, 1, 1};
	int my_array3_size = 5;

	int my_array4[] = {123, 123};
	int my_array4_size = 2;

	int my_array5[] = {1, 123, 123};
	int my_array5_size = 3;

	int my_array6[] = {123, 123, 166};
	int my_array6_size = 3;

	int my_array7[] = {1, 2, 8, 8 , 24, 60, 60, 60, 60, 75, 100, 100, 123};
	int my_array7_size = 13;


	my_array1_size = removeDuplicates(my_array1, my_array1_size);
	my_array2_size = removeDuplicates(my_array2, my_array2_size);
	my_array3_size = removeDuplicates(my_array3, my_array3_size);
	my_array4_size = removeDuplicates(my_array4, my_array4_size);
	my_array5_size = removeDuplicates(my_array5, my_array5_size);
	my_array6_size = removeDuplicates(my_array6, my_array6_size);
	my_array7_size = removeDuplicates(my_array7, my_array7_size);

	return(0);
}


// Function to remove the duplicates
int removeDuplicates(int a[], int array_size){
	int i, j;
	j = 0;
	// Print old array...
	printf("\n\nOLD : ");
	for(i = 0; i < array_size; i++){
		printf("[%d] ", a[i]);
	}

	// Remove the duplicates ...
	for (i = 1; i < array_size; i++){
		if (a[i] != a[j]){
			j++;
			a[j] = a[i]; // Move it to the front
		}
	}

	// The new array size..
	array_size = (j + 1);


	// Print new array...
	printf("\n\nNEW : ");
	for(i = 0; i < array_size; i++){
		printf("[%d] ", a[i]);
	}
	printf("\n\n");

	// Return the new size...
	return(j + 1);
}

File: leetcode/RemoveDuplicatesFromSortedArray/RemoveDuplicates.java
/*
Given a sorted array, remove the duplicates in place such that each element appear only once and return the new length.
Do not allocate extra space for another array, you must do this in place with constant memory.

For example,
Given input array nums = [1,1,2],

Your function should return length = 2, with the first two elements of nums being 1 and 2 respectively.
It doesn't matter what you leave beyond the new length.

idea:
very similar to Remove Element leetcode

note: always save first element (index = 0) to result array

i for all elements in A array, starting from 1
j for unique elements in A array,  starting from 0

for example
1 1 1 2 3 =>
1 2 1 2 3 =>
1 2 3 2 3

j controls 1 2 3
i controls 1 1 1 2 3 
*/

class RemoveDuplicates {
	public static void main(String[] args) {
		int my_array1[] = {1, 1, 2, 3, 5, 6, 6, 10, 25, 100, 123, 123};
		int my_array3[] = {1, 1, 1, 1, 1};
		int my_array5[] = {1, 123, 123};

		RemoveDuplicates eg = new RemoveDuplicates();
		System.out.println("New size of array1: " + my_array1.length + " " + eg.removeDuplicates(my_array1));
		System.out.println("New size of array3: " + my_array3.length + " " + eg.removeDuplicates(my_array3));
		System.out.println("New size of array5: " + my_array5.length + " " + eg.removeDuplicates(my_array5));
	}

	// Tue Jun 18 23:07:43 2019
	public int removeDuplicates(int[] nums) {
		// 记住 这个方法里 j 是 index 
		// 最后 j + 1
		int j = 0;
		
		int size = nums.length;
		
		for (int i = 1; i < size; i++) {
			int prev = nums[i - 1];
			int curr = nums[i];
			
			if (prev != curr) {
				nums[++j] = curr;
			}
		}
		
		return j + 1;
	}
	// similar as above, just while()
	public int removeDuplicates(int[] nums) {
		if (nums.length < 2) {
			return nums.length;
		}

		int i = 1;
		int j = 0;
		
		while (i < nums.length) {
			if (nums[j] != nums[i]) {
				j++;
				// this is to make sure all unique values go to left front
				nums[j] = nums[i];
			}
			i++;
		}
		
		return j + 1;
	}

	// best method
	public int removeDuplicates(int[] nums) {
		int j = 0;
		for (int i = 0; i < nums.length; i++) {
			if (nums[i] != nums[j]) {
				// note j still starts from 1
				nums[++j] = nums[i];
			}
		}

		return j + 1;
	}
	// method 2
	public int removeDuplicatesTwo(int a[]) {
		int i, j = 0;
		System.out.printf("\nOLD : ");
		for (i = 0; i < a.length; i++) {
			System.out.printf("[%d] ", a[i]);
		}
		// Remove the duplicates
		for (i = 1; i < a.length; i++) {
			if (a[i] != a[j]) {
				j++;
				a[j] = a[i]; // Move it to the front
			}
		}
		System.out.printf("\nNEW : ");
		for (i = 0; i < j + 1; i++) {
			System.out.printf("[%d] ", a[i]);
		}
		System.out.printf("\n");

		// The new array size, j + 1
		return j + 1;
	}
}
